js/Repeater/RepeaterPreviewService.js

```diff
@@ -0,0 +1,123 @@
+class RepeaterPreviewService {
+    /**
+     * Create / Update Repeater preview elements
+     * @param queryService {QueryService}
+     * @param activeFunctionService {ActiveFunctionService}
+     * @param inputValueService {InputValueService}
+     * @param options
+     */
+    constructor (
+        queryService,
+        activeFunctionService,
+        inputValueService,
+        options = {}
+    ) {
+        this.inputValueService = inputValueService;
+        this.queryService = queryService;
+        this.activeFunctionService = activeFunctionService;
+        this.emptyHTML = options.empty || 'empty';
+    }
+
+    /**
+     * Create a repeater preview
+     * @param state
+     * @param headings
+     * @param id
+     * @returns {HTMLElement}
+     */
+    create (state, headings, id) {
+        const previewRow = document.createElement('tr');
+
+        // For each heading, create a repeater preview
+        headings.forEach(heading => {
+            const name = heading.getAttribute(this.queryService.getAttr('preview-heading'));
+            const data = state[name];
+            let value = this.emptyHTML;
+
+            if (data === undefined) {
+                throw new Error(`The input "${name}" was not found in the Repeater.`);
+            }
+
+            const preview = document.createElement('td');
+
+            preview.setAttribute(this.queryService.getAttr('preview-update-id'), `${name}_${id}`);
+
+            data.value
+                .filter(input => input.selected && input.value)
+                .forEach(input => {
+                    value = this.print(input.ref, value, input.value);
+                });
+
+            preview.textContent = value;
+            previewRow.appendChild(preview);
+        });
+
+        return previewRow;
+    }
+
+    /**
+     * Update a repeater preview
+     * @param state
+     * @param headings
+     * @param root
+     * @param id
+     */
+    update (state, headings, root, id) {
+        headings.forEach(heading => {
+            const name = heading.getAttribute('data-repeater-for-name');
+            const data = state[name];
+            let value = this.emptyHTML;
+
+            // If our heading exists inside the state object
+            if (state[name]) {
+                const preview = root.querySelector(`[${this.queryService.getAttr('preview-update-id')}="${name}_${id}"]`);
+
+                // Set the value for each input in the state
+                data.value
+                    .filter(input => input.selected && input.value)
+                    .forEach(input => {
+                        value = this.print(input.ref, value, input.value);
+                    });
+
+                preview.innerText = value;
+            }
+        });
+    }
+
+    /**
+     * Toggle disabled state of preview UI, if a preview ID is not passed in
+     * we'll disabled all preview UI elements
+     * @param previewId?
```

Comment:
Why is the '?' needed here?

Reply:
That is the syntax in JSDOC for an optional function argument